Automatic track export silently fails

I was wondering why the automatic track export isn't working on the release version.
Later I connected the device to AS to get a logcat:

```
2020-11-11 20:22:38.992 26923-26923/? W/DocumentsContract: Failed to create document
    android.os.ParcelableException: java.io.FileNotFoundException: Failed to upload document with path /OpenTrack/308.kmz
        at android.os.ParcelableException$1.createFromParcel(ParcelableException.java:82)
        at android.os.ParcelableException$1.createFromParcel(ParcelableException.java:79)
        at android.os.Parcel.readParcelable(Parcel.java:2990)
        at android.os.Parcel.createException(Parcel.java:2083)
        at android.os.Parcel.readException(Parcel.java:2056)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:188)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140)
        at android.content.ContentProviderProxy.call(ContentProviderNative.java:658)
        at android.content.ContentResolver.call(ContentResolver.java:2049)
        at android.provider.DocumentsContract.createDocument(DocumentsContract.java:1333)
        at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:53)
        at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:45)
        at de.dennisguse.opentracks.util.ExportUtils.getExportDocumentFileUri(ExportUtils.java:88)
        at de.dennisguse.opentracks.util.ExportUtils.exportTrack(ExportUtils.java:38)
        at de.dennisguse.opentracks.util.ExportUtils.postWorkoutExport(ExportUtils.java:31)
        at de.dennisguse.opentracks.services.TrackRecordingService.endCurrentTrack(TrackRecordingService.java:434)
        at de.dennisguse.opentracks.services.TrackRecordingServiceBinder.endCurrentTrack(TrackRecordingServiceBinder.java:69)
        at de.dennisguse.opentracks.services.TrackRecordingServiceConnection.stopRecording(TrackRecordingServiceConnection.java:213)
        at de.dennisguse.opentracks.TrackRecordingActivity$4.onClick(TrackRecordingActivity.java:149)
        at de.dennisguse.opentracks.TrackController$ButtonDelay.lambda$run$1$TrackController$ButtonDelay(TrackController.java:161)
        at de.dennisguse.opentracks.-$$Lambda$TrackController$ButtonDelay$yzFjcE6Sa0ALJEX_mRxG_IsAjAI.run(Unknown Source:2)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:237)
        at android.app.ActivityThread.main(ActivityThread.java:8016)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1076)
     Caused by: java.io.FileNotFoundException: Failed to upload document with path /OpenTrack/308.kmz
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at android.os.ParcelableException.readFromParcel(ParcelableException.java:56)
        at android.os.ParcelableException$1.createFromParcel(ParcelableException.java:82) 
        at android.os.ParcelableException$1.createFromParcel(ParcelableException.java:79) 
        at android.os.Parcel.readParcelable(Parcel.java:2990) 
        at android.os.Parcel.createException(Parcel.java:2083) 
        at android.os.Parcel.readException(Parcel.java:2056) 
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:188) 
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140) 
        at android.content.ContentProviderProxy.call(ContentProviderNative.java:658) 
        at android.content.ContentResolver.call(ContentResolver.java:2049) 
        at android.provider.DocumentsContract.createDocument(DocumentsContract.java:1333) 
        at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:53) 
        at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:45) 
        at de.dennisguse.opentracks.util.ExportUtils.getExportDocumentFileUri(ExportUtils.java:88) 
        at de.dennisguse.opentracks.util.ExportUtils.exportTrack(ExportUtils.java:38) 
        at de.dennisguse.opentracks.util.ExportUtils.postWorkoutExport(ExportUtils.java:31) 
        at de.dennisguse.opentracks.services.TrackRecordingService.endCurrentTrack(TrackRecordingService.java:434) 
        at de.dennisguse.opentracks.services.TrackRecordingServiceBinder.endCurrentTrack(TrackRecordingServiceBinder.java:69) 
        at de.dennisguse.opentracks.services.TrackRecordingServiceConnection.stopRecording(TrackRecordingServiceConnection.java:213) 
        at de.dennisguse.opentracks.TrackRecordingActivity$4.onClick(TrackRecordingActivity.java:149) 
        at de.dennisguse.opentracks.TrackController$ButtonDelay.lambda$run$1$TrackController$ButtonDelay(TrackController.java:161) 
        at de.dennisguse.opentracks.-$$Lambda$TrackController$ButtonDelay$yzFjcE6Sa0ALJEX_mRxG_IsAjAI.run(Unknown Source:2) 
        at android.os.Handler.handleCallback(Handler.java:883) 
        at android.os.Handler.dispatchMessage(Handler.java:100) 
        at android.os.Looper.loop(Looper.java:237) 
        at android.app.ActivityThread.main(ActivityThread.java:8016) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1076) 
2020-11-11 20:22:38.993 26923-26923/? E/TrackRecordingServiceConnection: Unable to stop recording.
    java.lang.NullPointerException: Attempt to invoke virtual method 'android.net.Uri androidx.documentfile.provider.DocumentFile.getUri()' on a null object reference
        at de.dennisguse.opentracks.util.ExportUtils.getExportDocumentFileUri(ExportUtils.java:88)
        at de.dennisguse.opentracks.util.ExportUtils.exportTrack(ExportUtils.java:38)
        at de.dennisguse.opentracks.util.ExportUtils.postWorkoutExport(ExportUtils.java:31)
        at de.dennisguse.opentracks.services.TrackRecordingService.endCurrentTrack(TrackRecordingService.java:434)
        at de.dennisguse.opentracks.services.TrackRecordingServiceBinder.endCurrentTrack(TrackRecordingServiceBinder.java:69)
        at de.dennisguse.opentracks.services.TrackRecordingServiceConnection.stopRecording(TrackRecordingServiceConnection.java:213)
        at de.dennisguse.opentracks.TrackRecordingActivity$4.onClick(TrackRecordingActivity.java:149)
        at de.dennisguse.opentracks.TrackController$ButtonDelay.lambda$run$1$TrackController$ButtonDelay(TrackController.java:161)
        at de.dennisguse.opentracks.-$$Lambda$TrackController$ButtonDelay$yzFjcE6Sa0ALJEX_mRxG_IsAjAI.run(Unknown Source:2)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:237)
        at android.app.ActivityThread.main(ActivityThread.java:8016)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1076)
```

Looks like the URI is not valid anymore. But as a user I get no information that the automatic export didn't work. There should be a dialog to inform the user about this error or better open the directory chooser again to select a new directory.

Hmm, even choosing a new directory fails with the above error message.

Later in the logcat I see this Exception:

```
2020-11-11 20:28:58.969 17721-17721/? D/nodomain.freeyourgadget.gadgetbridge.externalevents.NotificationListener: Notification removed: de.dennisguse.opentracks: service
2020-11-11 20:28:58.969 5862-8527/? E/Parcel: Class not found when unmarshalling: de.dennisguse.opentracks.content.data.Track$Id
    java.lang.ClassNotFoundException: de.dennisguse.opentracks.content.data.Track$Id
        at java.lang.Class.classForName(Native Method)
        at java.lang.Class.forName(Class.java:454)
        at android.os.Parcel.readParcelableCreator(Parcel.java:3031)
        at android.os.Parcel.readParcelable(Parcel.java:2981)
        at android.os.Parcel.readValue(Parcel.java:2883)
        at android.os.Parcel.readArrayMapInternal(Parcel.java:3261)
        at android.os.BaseBundle.initializeFromParcelLocked(BaseBundle.java:292)
        at android.os.BaseBundle.unparcel(BaseBundle.java:236)
        at android.os.BaseBundle.getString(BaseBundle.java:1160)
        at android.content.Intent.getStringExtra(Intent.java:8548)
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:790)
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:694)
        at com.android.server.wm.ActivityStarter.startActivityMayWait(ActivityStarter.java:1925)
        at com.android.server.wm.ActivityStarter.execute(ActivityStarter.java:625)
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1707)
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1603)
        at com.android.server.wm.ActivityTaskManagerService.startActivity(ActivityTaskManagerService.java:1557)
        at android.app.IActivityTaskManager$Stub.onTransact(IActivityTaskManager.java:1690)
        at android.os.Binder.execTransactInternal(Binder.java:1021)
        at android.os.Binder.execTransact(Binder.java:994)
     Caused by: java.lang.ClassNotFoundException: de.dennisguse.opentracks.content.data.Track$Id
        at java.lang.Class.classForName(Native Method) 
        at java.lang.Class.forName(Class.java:454) 
        at android.os.Parcel.readParcelableCreator(Parcel.java:3031) 
        at android.os.Parcel.readParcelable(Parcel.java:2981) 
        at android.os.Parcel.readValue(Parcel.java:2883) 
        at android.os.Parcel.readArrayMapInternal(Parcel.java:3261) 
        at android.os.BaseBundle.initializeFromParcelLocked(BaseBundle.java:292) 
        at android.os.BaseBundle.unparcel(BaseBundle.java:236) 
        at android.os.BaseBundle.getString(BaseBundle.java:1160) 
        at android.content.Intent.getStringExtra(Intent.java:8548) 
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:790) 
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:694) 
        at com.android.server.wm.ActivityStarter.startActivityMayWait(ActivityStarter.java:1925) 
        at com.android.server.wm.ActivityStarter.execute(ActivityStarter.java:625) 
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1707) 
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1603) 
        at com.android.server.wm.ActivityTaskManagerService.startActivity(ActivityTaskManagerService.java:1557) 
        at android.app.IActivityTaskManager$Stub.onTransact(IActivityTaskManager.java:1690) 
        at android.os.Binder.execTransactInternal(Binder.java:1021) 
        at android.os.Binder.execTransact(Binder.java:994) 
2020-11-11 20:28:58.969 5862-8527/? W/Bundle: Failed to parse Bundle, but defusing quietly
    android.os.BadParcelableException: ClassNotFoundException when unmarshalling: de.dennisguse.opentracks.content.data.Track$Id
        at android.os.Parcel.readParcelableCreator(Parcel.java:3059)
        at android.os.Parcel.readParcelable(Parcel.java:2981)
        at android.os.Parcel.readValue(Parcel.java:2883)
        at android.os.Parcel.readArrayMapInternal(Parcel.java:3261)
        at android.os.BaseBundle.initializeFromParcelLocked(BaseBundle.java:292)
        at android.os.BaseBundle.unparcel(BaseBundle.java:236)
        at android.os.BaseBundle.getString(BaseBundle.java:1160)
        at android.content.Intent.getStringExtra(Intent.java:8548)
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:790)
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:694)
        at com.android.server.wm.ActivityStarter.startActivityMayWait(ActivityStarter.java:1925)
        at com.android.server.wm.ActivityStarter.execute(ActivityStarter.java:625)
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1707)
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1603)
        at com.android.server.wm.ActivityTaskManagerService.startActivity(ActivityTaskManagerService.java:1557)
        at android.app.IActivityTaskManager$Stub.onTransact(IActivityTaskManager.java:1690)
        at android.os.Binder.execTransactInternal(Binder.java:1021)
        at android.os.Binder.execTransact(Binder.java:994)
```

I did some more debugging. Automatic export after recording fails for NextCloud directory, but not for a local device directory. The NextCloud directory works when manually doing a "Export all".

The exception:

```
2020-11-15 21:24:59.472 3285-3285/de.dennisguse.opentracks.debug W/DocumentsContract: Failed to create document
    android.os.ParcelableException: java.io.FileNotFoundException: Failed to upload document with path /OpenTracksDebug/178.kmz
        at android.os.ParcelableException$1.createFromParcel(ParcelableException.java:82)
        at android.os.ParcelableException$1.createFromParcel(ParcelableException.java:79)
        at android.os.Parcel.readParcelable(Parcel.java:2990)
        at android.os.Parcel.createException(Parcel.java:2083)
        at android.os.Parcel.readException(Parcel.java:2056)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:188)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140)
        at android.content.ContentProviderProxy.call(ContentProviderNative.java:658)
        at android.content.ContentResolver.call(ContentResolver.java:2049)
        at android.provider.DocumentsContract.createDocument(DocumentsContract.java:1333)
        at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:53)
        at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:45)
        at de.dennisguse.opentracks.util.ExportUtils.getExportDocumentFileUri(ExportUtils.java:87)
        at de.dennisguse.opentracks.util.ExportUtils.exportTrack(ExportUtils.java:38)
        at de.dennisguse.opentracks.util.ExportUtils.postWorkoutExport(ExportUtils.java:31)
        at de.dennisguse.opentracks.services.TrackRecordingService.endCurrentTrack(TrackRecordingService.java:434)
        at de.dennisguse.opentracks.services.TrackRecordingServiceBinder.endCurrentTrack(TrackRecordingServiceBinder.java:69)
        at de.dennisguse.opentracks.services.TrackRecordingServiceConnection.stopRecording(TrackRecordingServiceConnection.java:213)
        at de.dennisguse.opentracks.TrackRecordingActivity$4.onClick(TrackRecordingActivity.java:149)
        at de.dennisguse.opentracks.TrackController$ButtonDelay.lambda$run$1$TrackController$ButtonDelay(TrackController.java:161)
        at de.dennisguse.opentracks.-$$Lambda$TrackController$ButtonDelay$yzFjcE6Sa0ALJEX_mRxG_IsAjAI.run(Unknown Source:2)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:237)
        at android.app.ActivityThread.main(ActivityThread.java:8016)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1076)
     Caused by: java.io.FileNotFoundException: Failed to upload document with path /OpenTracksDebug/178.kmz
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at android.os.ParcelableException.readFromParcel(ParcelableException.java:56)
        at android.os.ParcelableException$1.createFromParcel(ParcelableException.java:82) 
        at android.os.ParcelableException$1.createFromParcel(ParcelableException.java:79) 
        at android.os.Parcel.readParcelable(Parcel.java:2990) 
        at android.os.Parcel.createException(Parcel.java:2083) 
        at android.os.Parcel.readException(Parcel.java:2056) 
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:188) 
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140) 
        at android.content.ContentProviderProxy.call(ContentProviderNative.java:658) 
        at android.content.ContentResolver.call(ContentResolver.java:2049) 
        at android.provider.DocumentsContract.createDocument(DocumentsContract.java:1333) 
        at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:53) 
        at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:45) 
        at de.dennisguse.opentracks.util.ExportUtils.getExportDocumentFileUri(ExportUtils.java:87) 
        at de.dennisguse.opentracks.util.ExportUtils.exportTrack(ExportUtils.java:38) 
        at de.dennisguse.opentracks.util.ExportUtils.postWorkoutExport(ExportUtils.java:31) 
        at de.dennisguse.opentracks.services.TrackRecordingService.endCurrentTrack(TrackRecordingService.java:434) 
        at de.dennisguse.opentracks.services.TrackRecordingServiceBinder.endCurrentTrack(TrackRecordingServiceBinder.java:69) 
        at de.dennisguse.opentracks.services.TrackRecordingServiceConnection.stopRecording(TrackRecordingServiceConnection.java:213) 
        at de.dennisguse.opentracks.TrackRecordingActivity$4.onClick(TrackRecordingActivity.java:149) 
        at de.dennisguse.opentracks.TrackController$ButtonDelay.lambda$run$1$TrackController$ButtonDelay(TrackController.java:161) 
        at de.dennisguse.opentracks.-$$Lambda$TrackController$ButtonDelay$yzFjcE6Sa0ALJEX_mRxG_IsAjAI.run(Unknown Source:2) 
        at android.os.Handler.handleCallback(Handler.java:883) 
        at android.os.Handler.dispatchMessage(Handler.java:100) 
        at android.os.Looper.loop(Looper.java:237) 
        at android.app.ActivityThread.main(ActivityThread.java:8016) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1076) 

```

Don't know what happens here. 

The BadParcelableException while unmarshalling the Track.Id seems to be caused by the System Classloader, we should use the application Classloader. But I tried several things, like this:
- https://stackoverflow.com/questions/18126249/badparcelableexception-classnotfoundexception-when-unmarshalling-myclass-wh

But it didn't help. I even don't find the place where it happes. In the Stacktrace is no clue about the source or target activity.

--- a/src/main/java/de/dennisguse/opentracks/content/data/Track.java
+++ b/src/main/java/de/dennisguse/opentracks/content/data/Track.java
@@ -118,6 +118,10 @@ public class Track {
             this.id = id;
         }
 
+        protected Id(Parcel in) {
+            id = in.readLong();
+        }
+
         //TOOD Limit visibility to TrackRecordingService / ContentProvider
         public long getId() {
             return id;

--- a/src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java
+++ b/src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java
@@ -25,6 +25,8 @@ import android.content.SharedPreferences;
 import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
 import android.database.sqlite.SQLiteException;
 import android.net.Uri;
+import android.os.Bundle;
+import android.os.Handler;
 import android.os.IBinder;
 import android.os.PowerManager.WakeLock;
 import android.util.Log;
@@ -46,12 +48,14 @@ import de.dennisguse.opentracks.content.provider.ContentProviderUtils;
 import de.dennisguse.opentracks.content.provider.CustomContentProvider;
 import de.dennisguse.opentracks.content.provider.TrackPointIterator;
 import de.dennisguse.opentracks.content.sensor.SensorDataSet;
+import de.dennisguse.opentracks.io.file.exporter.ExportServiceResultReceiver;
 import de.dennisguse.opentracks.services.handlers.GpsStatusValue;
 import de.dennisguse.opentracks.services.handlers.HandlerServer;
 import de.dennisguse.opentracks.services.sensors.BluetoothRemoteSensorManager;
 import de.dennisguse.opentracks.services.sensors.ElevationSumManager;
 import de.dennisguse.opentracks.services.tasks.AnnouncementPeriodicTaskFactory;
 import de.dennisguse.opentracks.services.tasks.PeriodicTaskExecutor;
+import de.dennisguse.opentracks.settings.SettingsActivity;
 import de.dennisguse.opentracks.stats.TrackStatistics;
 import de.dennisguse.opentracks.stats.TrackStatisticsUpdater;
 import de.dennisguse.opentracks.util.ExportUtils;
@@ -69,7 +73,7 @@ import de.dennisguse.opentracks.util.TrackPointUtils;
  *
  * @author Leif Hendrik Wilden
  */
-public class TrackRecordingService extends Service implements HandlerServer.HandlerServerInterface {
+public class TrackRecordingService extends Service implements HandlerServer.HandlerServerInterface, ExportServiceResultReceiver.Receiver {
 
     private static final String TAG = TrackRecordingService.class.getSimpleName();
 
@@ -431,7 +435,7 @@ public class TrackRecordingService extends Service implements HandlerServer.Hand
             }
         }
 
-        ExportUtils.postWorkoutExport(this, track);
+        ExportUtils.postWorkoutExport(this, track, new ExportServiceResultReceiver(new Handler(), this));
 
         endRecording(true);
     }
@@ -751,4 +755,16 @@ public class TrackRecordingService extends Service implements HandlerServer.Hand
     public GpsStatusValue getGpsStatus() {
         return handlerServer.getGpsStatus();
     }
+
+    @Override
+    public void onReceiveResult(final int resultCode, final Bundle resultData) {
+        Log.w(TAG, "onReceiveResult: " + resultCode);
+        if (resultCode != ExportServiceResultReceiver.RESULT_CODE_SUCCESS) {
+            Intent intent = new Intent(this, SettingsActivity.class);
+            intent.putExtra(SettingsActivity.EXTRAS_CHECK_EXPORT_DIRECTORY, true);
+            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+            startActivity(intent);
+        }
+    }
+
 }

--- a/src/main/java/de/dennisguse/opentracks/settings/SettingsActivity.java
+++ b/src/main/java/de/dennisguse/opentracks/settings/SettingsActivity.java
@@ -1,10 +1,12 @@
 package de.dennisguse.opentracks.settings;
 
+import android.content.Intent;
 import android.content.SharedPreferences;
 import android.os.Bundle;
 import android.util.Log;
 import android.widget.Toast;
 
+import androidx.appcompat.app.AlertDialog;
 import androidx.appcompat.app.AppCompatActivity;
 import androidx.appcompat.widget.Toolbar;
 import androidx.documentfile.provider.DocumentFile;
@@ -32,8 +34,10 @@ import de.dennisguse.opentracks.util.StringUtils;
 public class SettingsActivity extends AppCompatActivity implements ChooseActivityTypeDialogFragment.ChooseActivityTypeCaller, ResetDialogPreference.ResetCallback {
 
     private static final String TAG = SettingsActivity.class.getSimpleName();
+    public static final String EXTRAS_CHECK_EXPORT_DIRECTORY = "Check Export Directory";
 
     private PrefsFragment prefsFragment;
+    private boolean checkExportDirectory = false;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
@@ -45,7 +49,27 @@ public class SettingsActivity extends AppCompatActivity implements ChooseActivit
         toolbar.setTitle(R.string.menu_settings);
         setSupportActionBar(toolbar);
 
+        Intent intent = getIntent();
+        if (intent != null && intent.hasExtra(EXTRAS_CHECK_EXPORT_DIRECTORY)) {
+            checkExportDirectory = true;
+        }
         onReset();
+    }
+
+    @Override
+    protected void onResume() {
+        if (checkExportDirectory) {
+            checkExportDirectory = false;
+            new AlertDialog.Builder(this)
+                    .setIcon(R.drawable.ic_logo_24dp)
+                    .setTitle(R.string.app_name)
+                    .setMessage(R.string.export_error_post_workout)
+                    .setNeutralButton(R.string.generic_ok, null)
+                    .create()
+                    .show();
+        }
+
+        super.onResume();
     }
 
     @Override

--- a/src/main/java/de/dennisguse/opentracks/util/ExportUtils.java
+++ b/src/main/java/de/dennisguse/opentracks/util/ExportUtils.java
@@ -17,17 +17,20 @@ import java.util.List;
 
 import de.dennisguse.opentracks.content.data.Track;
 import de.dennisguse.opentracks.io.file.TrackFileFormat;
+import de.dennisguse.opentracks.io.file.exporter.ExportService;
+import de.dennisguse.opentracks.io.file.exporter.ExportServiceResultReceiver;
 import de.dennisguse.opentracks.io.file.exporter.TrackExporter;
 
 public class ExportUtils {
 
     private static final String TAG = ExportUtils.class.getSimpleName();
 
-    public static void postWorkoutExport(Context context, Track track) {
+    public static void postWorkoutExport(Context context, Track track, ExportServiceResultReceiver resultReceiver) {
         if (PreferencesUtils.shouldInstantExportAfterWorkout(context)) {
             TrackFileFormat trackFileFormat = PreferencesUtils.getExportTrackFileFormat(context);
             DocumentFile directory = PreferencesUtils.getDefaultExportDirectoryUri(context);
 
+            ExportService.enqueue(context, resultReceiver, track.getId(), trackFileFormat, directory.getUri());
             exportTrack(context, trackFileFormat, directory, track);
         }
     }
@@ -36,6 +39,10 @@ public class ExportUtils {
         TrackExporter trackExporter = trackFileFormat.newTrackExporter(context);
 
         Uri exportDocumentFileUri = getExportDocumentFileUri(context, track.getId(), trackFileFormat, directory);
+        if (exportDocumentFileUri == null) {
+            Log.e(TAG, "Couldn't create document file for export");
+            return false;
+        }
 
         try (OutputStream outputStream = context.getContentResolver().openOutputStream(exportDocumentFileUri)) {
             if (trackExporter.writeTrack(track, outputStream)) {
@@ -80,7 +87,10 @@ public class ExportUtils {
         String exportFileName = getExportFileNameByTrackId(trackId, trackFileFormat.getExtension());
         Uri exportDocumentFileUri = findFile(context, directory.getUri(), exportFileName);
         if (exportDocumentFileUri == null) {
-            exportDocumentFileUri = directory.createFile(trackFileFormat.getMimeType(), exportFileName).getUri();
+            final DocumentFile file = directory.createFile(trackFileFormat.getMimeType(), exportFileName);
+            if (file != null) {
+                exportDocumentFileUri = file.getUri();
+            }
         }
         return exportDocumentFileUri;
     }

--- a/src/main/res/values/strings.xml
+++ b/src/main/res/values/strings.xml
@@ -212,6 +212,7 @@ limitations under the License.
     <string name="export_with_photos">with photos</string>
     <string name="export_without_photos">without photos</string>
     <string name="export_error">Exported %1$d of %2$s to %3$s</string>
+    <string name="export_error_post_workout">Post-workout export failed, please check the export directory.</string>
     <string name="export_option">as %1$s to %2$s</string>
     <string name="export_progress_message">Exporting to %1$s&#8230;</string>
     <string name="export_success">Exported %1$s to %2$s</string>

